Systemd stage 1 networkd

ElvishJerricco opened the pull request:
###### Description of changes

Use systemd-networkd for networking in stage 1.

- Adds a mechanism for users/groups in initrd, since that's required for the `systemd-network` user/group.
- Configures `networking.interfaces` to the same level as scripted initrd.
- Supports SSH and OpenVPN
- Adds dbus with socket activation so that `networkctl` can work.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ElvishJerricco committed:
systemd-initrd: Don't use SYSTEMD_SULOGIN_FORCE

flokli commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -69,18 +69,10 @@ let
     "systemd-sysctl.service"
     "systemd-tmpfiles-setup-dev.service"
-    "systemd-tmpfiles-setup.service"
+    # This imposes an ordering the prevents remote access during disk unlock
```
```suggestion
    # This imposes an ordering that prevents remote access during disk unlock
```

flokli commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -364,7 +356,7 @@ in {
 
         "/etc/systemd/system.conf".text = ''
           [Manager]
-          DefaultEnvironment=PATH=/bin:/sbin ${optionalString (isBool cfg.emergencyAccess && cfg.emergencyAccess) "SYSTEMD_SULOGIN_FORCE=1"}
```
This is unrelated to this PR, no?

ElvishJerricco replied:
Sort of? I needed it because sshd requires the root account to be unlocked.

flokli commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -1941,13 +1928,7 @@ in
-          config.environment.etc."systemd/networkd.conf".source
-        ];
-      };
+      systemd.services.systemd-networkd.aliases = [ "dbus-org.freedesktop.network1.service" ];
```
Do we also have dbus in stage1? :eyes:

arianvp replied:
We do not. But networkd runs fine without it

flokli replied:
So should this alias not be in here? Or should it be? Or it doesn't matter?

flokli replied:
We should probably at least have a comment explaining this line. I was quite surprised to see dbus stuff in stage1.

ElvishJerricco replied:
I was trying to leave as much code in common as the stage 1 systemd options currently support. No rhyme or reason for this in particular to be in stage 1.

flokli commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -1969,6 +1950,32 @@ in
+    })
+  ];
+
+  stage2Config = mkMerge [
```
What is stage 2 here? I assumed stage 2 was this other bash script that we have, that invokes the real system (and we might be able to absorb everything into a systemd-powered stage 1…)

ElvishJerricco replied:
Stage 2 just refers to the whole OS after we leave initrd, as I understand the term.

flokli reviewed (commented):
I don't quite understand why we can't just also ship all `.network` and `.netdev` files in the initrd. Did you run into troubles when doing that?

ElvishJerricco committed:
systemd-initrd: networkd

ElvishJerricco committed:
systemd-initrd: Fix up root directory mode

ElvishJerricco committed:
systemd-initrd: Add users and groups with static IDs.

SuperSandro2000 commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -12,16 +12,7 @@ let
-  '';
-
-  cfg  = config.services.openssh;
+  _cfg  = config.services.openssh;
   cfgc = config.programs.ssh;
```
Can we name those better?

SuperSandro2000 commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -422,15 +398,8 @@ in
-            restartTriggers = optionals (!cfg.startWhenNeeded) [
-              config.environment.etc."ssh/sshd_config".source
-            ];
+            path = [ cfgc.package pkgs.gawk pkgs.coreutils ];
```
```suggestion
            path = with pkgs; [ cfgc.package gawk coreutils ];
```

ElvishJerricco replied:
I'd rather avoid `with` whenever reasonable. Even in simple examples like this.

SuperSandro2000 replied:
I actually would say we should highly recommend it to be used in small lists with repeating prefixes. With my change it is very easy to see with a glance what packages are added but with the extra pkgs. prefix I can barely parse it.

ElvishJerricco replied:
On top of the fact that I and many others strongly disagree with that on account of `with` introducing dynamic scoping that requires logical analysis instead of simpler lexical analysis to understand, this is also not a matter of substance to this PR as the line being commented on is a pre-existing one.

SuperSandro2000 commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -561,6 +520,53 @@ in
+      UsePAM yes
+    '';
+
+    users.users.sshd =
+      {
+        isSystemUser = true;
+        group = "sshd";
+        description = "SSH privilege separation user";
+      };
```
```suggestion
    users.users.sshd = {
      isSystemUser = true;
      group = "sshd";
      description = "SSH privilege separation user";
    };
```

SuperSandro2000 commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -561,6 +520,53 @@ in
+      };
+    users.groups.sshd = {};
+
+    networking.firewall.allowedTCPPorts = if cfg.openFirewall then cfg.ports else [];
```
```suggestion
    networking.firewall.allowedTCPPorts = lib.optional cfg.openFirewall cfg.ports;
```

SuperSandro2000 commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -561,6 +520,53 @@ in
+
+    networking.firewall.allowedTCPPorts = if cfg.openFirewall then cfg.ports else [];
+
+    security.pam.services.sshd =
+      { startSession = true;
+        showMotd = true;
+        unixAuth = cfg.passwordAuthentication;
+      };
```
```suggestion
    security.pam.services.sshd = {
      startSession = true;
      showMotd = true;
      unixAuth = cfg.passwordAuthentication;
    };
```

SuperSandro2000 commented on nixos/modules/system/boot/initrd-ssh.nix:
```diff
@@ -33,8 +35,9 @@ in
-      default = "/bin/ash";
+      type = types.nullOr types.str;
+      default = null;
+      defaultText = ''"/bin/ash"'';
```
```suggestion
      defaultText = "/bin/ash";
```
I don't think we want the quotes in the final string.

ElvishJerricco replied:
Yes we do. The default text is supposed to be a valid expression for the option.

SuperSandro2000 commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -1911,12 +1907,13 @@ in
   };
 
-  config = mkMerge [
+  commonConfig = config: let cfg = config.systemd.network; in mkMerge [
```
Normally does are defined at the very top but since the file is already very big I am a bit unsure here.

ElvishJerricco replied:
@SuperSandro2000 `commonConfig` is an abstraction over the same options in both stage 1 and stage 2. They need different configs, so it has to be a function argument.

SuperSandro2000 commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -135,6 +127,13 @@ let
 
 in {
+  options.boot.initrd.environment.etc = mkOption {
+    description = "Set of files that have to be linked in /etc";
+    visible = false;
+    default = {};
+    type = options.environment.etc.type;
```
```suggestion
    type = options.environment.etc.type;
    visible = false;
    default = {};
    description = "Set of files that have to be linked in /etc";
```

ElvishJerricco replied:
This is going to need some tweaking anyway. The `type = options.environment.etc.type` trick is dirty and for some reason it doesn't work when evaluating the manual.

SuperSandro2000 commented on pkgs/build-support/kernel/make-initrd-ng.nix:
```diff
@@ -77,11 +77,12 @@ in
   mkdir ./root
   make-initrd-ng "$contentsPath" ./root
   mkdir "$out"
+  mkdir -p ./root/var/empty
```
```suggestion
  mkdir -p ./root/var/empty
  make-initrd-ng "$contentsPath" ./root
  mkdir "$out"
```

ElvishJerricco commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -135,6 +135,13 @@ let
+  options.boot.initrd.environment.etc = mkOption {
+    visible = false;
+    default = {};
+    type = options.environment.etc.type;
```
Before merging: How should we share this type, if at all? This way does not work when building the nixos manual, though it was awfully convenient for developing the feature.

ElvishJerricco commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -1982,6 +1988,62 @@ in
+      systemd.targets.network.wantedBy = ["emergency.target"];
+
+      systemd.contents = {
+        "/etc/systemd/networkd.conf" = renderConfig config.boot.initrd.systemd.network.config;
```
I split this between the stage 1 and 2 configs before I had the environment.etc commit. I should probably re-unify them in the commonConfig.

ElvishJerricco committed:
systemd-initrd: sshd

ElvishJerricco commented:
I've added a warning if no networks are configured, since we don't automatically configure anything like the scripted initrd does.

I've also decided not to mangle `sshd.nix` to have it serve both stage 1 and 2. Instead I've made a much simpler patch to implement effectively what the scripted initrd does with a simple service.

ElvishJerricco committed:
systemd-initrd: Support secrets when boot loader doesn't

initrd-secrets: Fix service config with systemd-stage-1

colemickens commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -1979,7 +1962,30 @@ in
+
+      systemd.services.systemd-networkd.aliases = [ "dbus-org.freedesktop.network1.service" ];
 
       networking.iproute2 = mkIf (cfg.config.addRouteTablesToIPRoute2 && cfg.config.routeTables != { }) {
```
It seems like `cfg` doesn't exist here? I get an error when I try to use this:

```
error: undefined variable 'cfg'
         1994|
         1995|       networking.iproute2 = mkIf (cfg.config.addRouteTablesToIPRoute2 && cfg.config.routeTables != { }) {
             |                                   ^
         1996|         enable = mkDefault true;
```

ElvishJerricco replied:
Ah, I hadn't been running the `nixos/tests/systemd-networkd.nix` test. Will do that going forward.

ElvishJerricco commented:
Ok, I've added tests for networking and ssh. I think the last two things to do before taking this out of WIP are:

- [ ] OpenVPN support
- [ ] Decide whether/how to support automatically configuring interfaces like we do with the scripted initrd.

ElvishJerricco committed:
systemd-initrd: Automatically configure networking.interfaces

ElvishJerricco commented:
Aside from openvpn, the only other thing not implemented is `boot.initrd.network.flushBeforeStage2`, which defaults to `true`. Now, I can see the value in this, and I understand why it was made the default originally, but `systemd-networkd` has no way of bringing every interface down on service stop (as far as I can see; please double check me). Upstream seems to consider it desirable to keep interfaces configured across stages. Can I just throw in a warning or something about this and call it a day? Or do we want an actual solution for this option, especially considering the present version of this branch is in conflict with the default of the nixos option?

gkleen commented:
> […] Can I just throw in a warning or something about this and call it a day? Or do we want an actual solution for this option, especially considering the present version of this branch is in conflict with the default of the nixos option?

Fwiw I manage a system where there is a separate interface for "management traffic" (LUKS pw) during stage 1 that's brought down before switching to stage 2.  
I would certainly like not having to plug in a bespoke systemd service in stage 1.

Then again I don't really see a way to implement an "actual solution" – networkd seems to just have no provisions to *ever* bring an interface down.  
Maybe throw that warning but also provide an option for a less-bespoke systemd service in stage 1 to bring a given list of interfaces down?

flokli commented:
> > […] Can I just throw in a warning or something about this and call it a day? Or do we want an actual solution for this option, especially considering the present version of this branch is in conflict with the default of the nixos option?
> 
> Fwiw I manage a system where there is a separate interface for "management traffic" (LUKS pw) during stage 1 that's brought down before switching to stage 2. I would certainly like not having to plug in a bespoke systemd service in stage 1.

I think it should be fine to throw a warning, and document how to add a small systemd unit to initrd to bring the interface down. It's an uncommon enough usecase.

ElvishJerricco committed:
systemd-initrd: OpenVPN

ElvishJerricco committed:
systemd-initrd: Flush networkd

ElvishJerricco committed:
systemd-initrd: dbus

ElvishJerricco commented:
I've pushed a new rebase including both OpenVPN and `flushBeforeStage2` support. I think this is ready for review.

ElvishJerricco commented:
Hm, the `dbus` addition increases the initrd size by roughly 700K and is only required for the `flushBeforeStage2` support... That should probably be optional.

ElvishJerricco commented:
Ok, `dbus` is now disabled by default and enabled when initrd networking is enabled. Also, I added tests to make sure `flushBeforeStage2` actually works.

ElvishJerricco commented on nixos/modules/system/boot/initrd-openvpn.nix:
```diff
@@ -71,6 +78,15 @@ in
+      wantedBy = ["initrd.target"];
+      path = [pkgs.iproute2];
+      after = ["network.target" "initrd-nixos-copy-secrets.service"];
+      serviceConfig.ExecStartPre = "/bin/mkdir /tmp";
```
This will need to be changed once staging reaches master because of https://github.com/NixOS/nixpkgs/pull/216826/files#diff-68fef76b66b99ee77baba62969a434c1f912c2fcf8765bfc884b7aa5b0e70d35R375

ElvishJerricco commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -2077,6 +2077,26 @@ let
+      systemd.services.nixos-flush-networkd = mkIf config.boot.initrd.network.flushBeforeStage2 {
+        description = "Flush Network Configuration";
+        wantedBy = ["initrd.target"];
+        after = ["systemd-networkd.service" "dbus.socket" "dbus.service"];
```
It's *possible* we should have an ordering between this service and `initrd-switch-root.target`. When there's any ordering at all between two units, and a transaction requires one to be stopped and one to be started (such as `systemctl isolate initrd-switch-root.target`), the stopping units are always stopped before the starting units are started. But that's only true if there's an ordering between them. Otherwise the stopping and starting may be simultaneous, which *I think* is bad in this case?

ElvishJerricco replied:
@oddlama whether or not the unit is stopped isn't the issue. The issue is whether it's stopped *before* `initrd-switch-root.target` is reached. When systemd isolates a unit, it *does* take the stop action on all units that aren't wanted by the unit being isolated, but that doesn't mean they'll be stopped *first*. They may end up being stopped after the unit being isolated has been started.

oddlama replied:
I see. So i dug a little deeper and it seems you are absolutely right, this would execute in parallel if no explicit dependencies are defined. There's a systemd issue about a related topic which [talks about this a bit in (and following comment)](https://github.com/systemd/systemd/issues/3436#issuecomment-224503276). Not sure whether we should also think about the `DefaultDependencies=no` stuff mentioned there, but from what I read adding an order would definitely be a good idea.

oddlama commented on pkgs/build-support/kernel/make-initrd-ng.nix:
```diff
@@ -78,14 +78,14 @@ in
     cat $PREP >> $out/initrd
   done
-  (cd root && find * .[^.*] -print0 | sort -z | cpio -o -H newc -R +0:+0 --reproducible --null | eval -- $compress >> "$out/initrd")
+  (cd root && find . -print0 | sort -z | cpio -o -H newc -R +0:+0 --reproducible --null | eval -- $compress >> "$out/initrd")
```
This looks like a simplification to me so I'm just wondering whether it is intended that the new command includes an extra output `.` compared to the replaced command (no `-mindepth 1`).

Also [this](https://unix.stackexchange.com/questions/453127) might be an interesting related note. Not sure if applicable here.

ElvishJerricco replied:
> wondering whether it is intended that the new command includes an extra output `.` compared to the replaced command

Including the root directory in the `find` output is important because it ensures it has the proper mode in the initrd.

> Also [this](https://unix.stackexchange.com/questions/453127) might be an interesting related note. Not sure if applicable here.

I don't believe that's applicable. We aren't unpacking this archive at all; the kernel just reads it into memory at startup. Plus one of the comments in that thread suggests most cpio implementations don't have this problem, including the ones installed on NixOS and macOS.

ElvishJerricco commented on nixos/modules/system/boot/initrd-ssh.nix:
```diff
@@ -177,8 +189,8 @@ in
-      echo '${cfg.shell}' > /etc/shells
-      echo 'root:x:0:0:root:/root:${cfg.shell}' > /etc/passwd
+      echo '${shell}' > /etc/shells
+      echo 'root:x:0:0:root:/root:${shell}' > /etc/passwd
```
This code is for the scripted initrd, not the systemd initrd. It isn't run with systemd initrd enabled. It's there to set up the `/etc/shells` and `/etc/passwd` files for the scripted initrd. The reason it appears in the diff of this PR is that the `shell` value has been abstracted out into a variable; `boot.initrd.network.shell` is not supported by the systemd initrd in this PR, so the value has been abstracted out into a variable to enable a clear error message about it.

ElvishJerricco commented on nixos/modules/tasks/network-interfaces-systemd.nix:
```diff
@@ -28,9 +28,160 @@ let
+      # of 1024, so that Ethernet is preferred if both are
+      # available.
+      dhcpV4Config.RouteMetric = 1025;
+      ipv6AcceptRAConfig.RouteMetric = 1025;
```
Yea as you noted, this code was moved, not added. If this should be changed, let's do that in a different PR.

mweinelt commented on nixos/lib/systemd-types.nix:
```diff
@@ -40,7 +40,7 @@ rec {
 
       target = mkOption {
-        type = types.path;
+        type = types.str;
```
If I remember correctly `types.path` also accepts strings that represent paths. So this generalizes what is already accepted.

ElvishJerricco replied:
@mweinelt the issue with `types.path` was that it doesn't work for `boot.initrd.environment.etc`, since it requires the first character to be `/`. Though, as I said before, I'm thinking we can probably get rid of `boot.initrd.environment.etc` because I bared even used it in this PR and it's easy to deal without.

mweinelt commented on nixos/tests/systemd-initrd-networkd.nix:
```diff
@@ -0,0 +1,45 @@
+      boot.initrd.systemd = {
+        enable = true;
+        network.networks."99-eth0" = {
+          matchConfig.Name = "eth0";
```
Will we have predicitve interface naming in the initrd?

ElvishJerricco replied:
This was removed in a later commit in the commit titled `systemd-initrd: Automatically configure networking.interfaces`. It should be handled exactly the same way as stage 2 does it. Is there more that needs to be done for predictable interface naming?

mweinelt replied:
Predictable names come from udev, so if udev is around to do it, probably not.

mweinelt commented on nixos/modules/system/boot/initrd-ssh.nix:
```diff
@@ -119,9 +124,11 @@ in
       Port ${toString cfg.port}
 
       PasswordAuthentication no
+      AuthorizedKeysFile %h/.ssh/authorized_keys %h/.ssh/authorized_keys2 /etc/ssh/authorized_keys.d/%u
```
Let's keep it around, I don't expect it to just be removed after it has been kept for 20 years.

dasJ replied:
Created a PR for the stage 2 config: #226010

mlyxshi commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -359,6 +405,45 @@ in {
+      "passwd".text = ''
+        ${lib.concatStringsSep "\n" (lib.mapAttrsToList (n: { uid, group }: let
+          g = cfg.groups.${group};
+        in "${n}:x:${toString uid}:${toString g.gid}:::") cfg.users)}
```
The emergency shell complains `Could not chdir to home directory : No such file or directory, Logging in with home = "/"`
Set root account home to  '/root', and other accounts to '/var/empty'?

phaer commented on nixos/modules/system/boot/initrd-openvpn.nix:
```diff
@@ -59,6 +59,13 @@ in
+      "${pkgs.openvpn}/bin/openvpn"
+      "${pkgs.iproute2}/bin/ip"
+      "${pkgs.glibc}/lib/libresolv.so.2"
+      "${pkgs.glibc}/lib/libnss_dns.so.2"
```
I see the space-constrains, but having "ip" around for debugging might be useful for emergency shells?

And isn't `libnss_dns.so.2` required for things such as sshds UseDNS as well?

dasJ replied:
About the first item: Why? I can trivially add it if I need it and it keeps the default size low.

About the second item: I see that but why not add it to the sshd module then? Or does ovpn use it as well?

phaer commented:
> I think this PR has been open for way too long ;-)

I'd love to see this merged as well and am happy to help. Maybe the PRs one-year-birthday on April 18 would be a nice occasion for a merge ;-)

ElvishJerricco commented:
I'm pushing some final changes.

1. I've removed `boot.initrd.environment.etc`. It wasn't useful enough to justify, and it risked misleading people into thinking it would work for scripted stage 1.
2. I've added tests for systemd stage 1 to `nixos/tests/predictable-interface-names.nix`. This revealed the fact that `99-default.link` should be in systemd stage 1 regardless of whether networking is enabled, so it's been moved to the udev module.
3. I've added unit ordering details to `nixos-flush-networkd.service` for `initrd-switch-root.target` and `shutdown.target`.
4. I've uncommented `systemd-tmpfiles-setup.service`, as that *does* seem to work in the current state of this branch. I originally commented it out when I was trying to share the `sshd` module between stages 1 and 2, and there was an ordering problem with it and LUKS disks. That's no longer the case.
5. I've reworked the `systemd-initrd-networkd-ssh` test so that it tests unlocking a LUKS disk, rather than a random `systemd-ask-passphrase` query. This is how I verified that `systemd-tmpfiles-setup` wasn't creating the ordering problem that it had been creating before. It's also just a more typical use case.
6. I set users' home directories to `/var/empty` to address @mlyxshi's comment.

And with that, I don't think I have any more concerns with this. Although I'd like to get an approving review from someone on @NixOS/systemd who knows a bit about systemd stage 1 and/or networkd first, if anyone wants to hit the merge button I wouldn't object.

0x4A6F commented on nixos/tests/predictable-interface-names.nix:
```diff
@@ -8,25 +8,48 @@ let
+
+      boot.initrd.systemd = {
+        enable = systemdStage1;
+        initrdBin = [ pkgs.iproute ];
```
Had to change occurences of `iproute` with `iproute2` to be able to run the tests:
```diff
- pkgs.iproute
+ pkgs.iproute2
```

ElvishJerricco replied:
But those are the same derivation?

colemickens replied:
I think it's because they have `allowAliases = false;` in this `nixpkgs.config`. (I think this because I hit issues like this too sometimes with that option set, and the is a corresponding line in `pkgs/top-level/aliases.nix`.

ElvishJerricco replied:
Ok, I changed all uses to `iproute2`

ElvishJerricco commented:
Rebased on nixos-unstable-small, along with the openvpn `mkdir /tmp` change mentioned above

flokli commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -3181,6 +3181,29 @@ let
+          # This service does nothing when starting, but brings down
+          # interfaces when switching root. This is the easiest way to
+          # ensure proper ordering while stopping
+          Type = "oneshot";
```
Can you link to the docs describing this should be flushed like this? I'm also surprised we need to manually parse the output and networkctl down individually - is there no command to do this better? Also, what about network mounts set up during initrd?

ElvishJerricco replied:
`networkctl` does not have any command / option for bringing down all interfaces, unfortunately.

ElvishJerricco replied:
Updated comment with reference to systemd documentation on the matter.

@flokli As for network mount units, that's already broken if you use `flushBeforeStage2` with scripted initrd. With systemd initrd, `flushBeforeStage2` is disabled by default though. And yes, I should think you would expect a network file system to break if you enable the option that brings down networking :P

flokli replied:
I meant docs on the "flushing networkd" part, and that there's no command for it ;-)

ElvishJerricco replied:
ah

ElvishJerricco replied:
Oh, then I'm not sure what you want linked? The `flushBeforeStage2` option means "I want interfaces brought down before stage 2", so `networkctl down` is how you do that. Not really anything to link to. I can leave a comment mentioning that there's no `--all` argument or anything like that though.

flokli replied:
> Oh, then I'm not sure what you want linked?

Documentation about why we need to do the manual flushing in first place ;-)

I looked at https://github.com/systemd/systemd/issues/780#issuecomment-166148078, which suggests flushing has been implemented in networkd, and there's even possibilities to have it not touch certain interfaces (`CriticalInterface`)

gkleen replied:
> I looked at [systemd/systemd#780 (comment)](https://github.com/systemd/systemd/issues/780#issuecomment-166148078), which suggests flushing has been implemented in networkd, and there's even possibilities to have it not touch certain interfaces (`CriticalInterface`)

@flokli reads to me like the discussion over there is about deconfiguring addresses when networkd *first takes possession* of an interface, not about any behaviour wrt. to shutdown.  
They don't seem to use the word “flush” in the same sense as it is used here.

ElvishJerricco replied:
Comment added. @flokli Does that answer the question?

flokli replied:
Sorry, i still dont understand why it's desirable to flush, and why this is on by default.

If i use network in initrd, I'd expect systemd-networkd in initrd config to be semi-compatible with the one on my real system, so i don't quite understand what we gain from flushing all addresses, and this is not very obvious from reading the option description. What's the actual breakage for a "default user" (probably doing DHCP on all wired interfaces) if this is left on?

Also note this only brings interfaces down (and maybe removes addresses or links)? It doesn't delete stacked interfaces like vlans or bonds.

ElvishJerricco replied:
> Sorry, i still dont understand why it's desirable to flush, and why this is on by default.

@flokli Well, for one thing, this PR changes it so it *isn't* on by default if systemd initrd is enabled. And whether or not the desire to flush makes sense is sort of immaterial to this PR, no? The option has existed in the scripted initrd for a long time. This PR merely implements the feature, and disables it by default with systemd initrd because you're right, it doesn't really make a whole lot of sense most of the time I don't think

mweinelt replied:
> Sorry, i still dont understand why it's desirable to flush, and why this is on by default.

Because the stage2 may want to do things with the interfaces, that it cannot do, while they're still up, like for example renaming them or setting other netdev properties.

flokli replied:
I'd have assumed renaming happens by `.link` units (parsed by udev), way before stage2 needs to do it. See also 355c58e485e76a4b91bb4613ee2a3e1bae809d28.

What's it that you can't express with `.link` files? Maybe it could become part of the description, so it's more understandable where this is necessary.

In any case, as this flushing is disabled by default, and opt-in, it's fine now IMHO.

flokli reviewed (commented):
I'd personally feel more comfortable if the OpenVPN and flush bits would be separate PRs. It might trigger some more discussion, blocking this PR more than necessary.

ElvishJerricco commented on nixos/modules/system/boot/initrd-openvpn.nix:
```diff
@@ -71,6 +78,14 @@ in
     boot.initrd.network.postCommands = ''
```
The `network.postCommands` part is for scripted stage 1. The systemd stage 1 has been deliberately designed not to implement the `boot.initrd.*Commands` options, for reasons outlined in the original PR.

ElvishJerricco replied:
I've pushed changes adopting the convention of using `mkIf (!config.boot.initrd.systemd.enable)` for these `*Commands` options in the ssh and openvpn modules, in order to make it clear what they're for, and to help ensure those options are not used at all when systemd initrd is in use.

dasJ commented on nixos/modules/system/boot/initrd-ssh.nix:
```diff
@@ -217,6 +229,36 @@ in
+
+    # Systemd initrd stuff
+    boot.initrd.systemd = mkIf config.boot.initrd.systemd.enable {
+      users.sshd = { uid = 1; group = "sshd"; };
```
Sidenote: do we assert that these UIDs are unique somewhere?

lilyinstarlight replied:
I don't think this does anywhere and it definitely should. Also wish we didn't have to hardcode uid `1` here, but I see why we probably do

ElvishJerricco replied:
It doesn't assert that anywhere... FWIW we're hopefully only ever using the implicit `users.foo = {}` style, which inherits IDs from the stage 2 config, which I believe does have such an assertion. `sshd` is the unfortunate exception to this. Anyway, I suppose it's worth asserting uniqueness

ElvishJerricco replied:
The assertion was actually extremely easy to add by moving the code for this to the `users-groups.nix` module, which is arguably a more appropriate location anyway.

dasJ commented on nixos/modules/system/boot/initrd-ssh.nix:
```diff
@@ -217,6 +229,36 @@ in
+        unitConfig.DefaultDependencies = false;
+        serviceConfig = {
+          ExecStart = "${package}/bin/sshd -D -f /etc/ssh/sshd_config";
+          Type = "simple";
```
That's the default. I'd rather see something along the lines of:
```
KillMode=process
Restart=on-failure
```

ElvishJerricco replied:
I find it better to be explicit with `Type` anyway, if only to make it obvious at a glance what it is.

@dasJ Can you explain what that would do? Sounds like it would keep ssh sessions alive when ssh is killed, and sshd would restart whenever it fails? I guess that makes sense.

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -445,6 +493,31 @@ in {
       ] ++ jobScripts;
 
+      users = {
+        root = {};
```
Wouldn't it be a good idea to force UID 0 here? Just in case

lilyinstarlight replied:
It pulls uid from the global config by default, so it should already be set to 0 from [`misc/ids.nix`](https://github.com/NixOS/nixpkgs/blob/3fb8eedc450286d5092e4953118212fa21091b3b/nixos/modules/misc/ids.nix#L42) and [`config/users-groups.nix`](https://github.com/NixOS/nixpkgs/blob/3fb8eedc450286d5092e4953118212fa21091b3b/nixos/modules/config/users-groups.nix#L548), right?

ElvishJerricco replied:
Yea it'll get it from stage 2, and I'm comfortable leaving that as is. If that ever gets messed up, we're in for all sorts of hurt anyway.

lilyinstarlight commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -3111,8 +3090,38 @@ in
+        ];
+      };
+
+      systemd.services.systemd-networkd.aliases = [ "dbus-org.freedesktop.network1.service" ];
```
Was there a reason this `aliases` was moved out of the attrset right above it? I mean it should work either way I think, but wasn't sure if there was some motivation to do this

ElvishJerricco replied:
Oh, I guess that's just a byproduct of organically evolving this over a year :P I should move that.

lilyinstarlight commented on nixos/modules/tasks/network-interfaces-systemd.nix:
```diff
@@ -28,9 +28,160 @@ let
 in
 
 {
+  imports = [(mkIf config.boot.initrd.network.enable {
+    # Note this is if initrd.network.enable, not if
+    # initrd.systemd.network.enable. By setting the latter and not the
+    # former, the user retains full control over the configuration.
+    boot.initrd.systemd.network = mkMerge [(genericDhcpNetworks true) interfaceNetworks];
+  })];
```
Why is this done as an import?

ElvishJerricco replied:
Pretty much exclusively to avoid re-indenting the entire file. The `config` definition is within a `mkIf` that doesn't apply here. I would actually very much appreciate a suggestion on how to do this better. If others think I ought to just do the reindentation anyway, we can do that.

lilyinstarlight replied:
Oof, I see what you mean now. I'd personally rather a reindentation than a `config`-depdendent `imports` hack, but I will defer to what others think since I don't feel all that strongly on it

A cheap trick you could do to avoid reindenting the file (since it's already a level down due to having `config` instead of just using the top-level attrset for config) is to change it like this:

```diff
diff --git a/nixos/modules/tasks/network-interfaces-systemd.nix b/nixos/modules/tasks/network-interfaces-systemd.nix
index 6a811ff5872..6ab1c5d552b 100644
--- a/nixos/modules/tasks/network-interfaces-systemd.nix
+++ b/nixos/modules/tasks/network-interfaces-systemd.nix
@@ -175,15 +175,16 @@ let
 
 in
 
-{
-  imports = [(mkIf config.boot.initrd.network.enable {
+mkMerge [
+
+  (mkIf config.boot.initrd.network.enable {
     # Note this is if initrd.network.enable, not if
     # initrd.systemd.network.enable. By setting the latter and not the
     # former, the user retains full control over the configuration.
     boot.initrd.systemd.network = mkMerge [(genericDhcpNetworks true) interfaceNetworks];
-  })];
+  })
 
-  config = mkIf cfg.useNetworkd {
+  (mkIf cfg.useNetworkd {
 
     assertions = [ {
       assertion = cfg.defaultGatewayWindowSize == null;
@@ -450,6 +451,6 @@ in
               bindsTo = [ "systemd-networkd.service" ];
           };
       };
-  };
+  })
 
-}
+]
```

But idk how much better that really is

ElvishJerricco replied:
@lilyinstarlight How about this?

```diff
diff --git a/nixos/modules/tasks/network-interfaces-systemd.nix b/nixos/modules/tasks/network-interfaces-systemd.nix
index 6a811ff5872..0fcd3c10219 100644
--- a/nixos/modules/tasks/network-interfaces-systemd.nix
+++ b/nixos/modules/tasks/network-interfaces-systemd.nix
@@ -176,14 +176,16 @@ let
 in
 
 {
-  imports = [(mkIf config.boot.initrd.network.enable {
+  config = mkMerge [
+
+  (mkIf config.boot.initrd.network.enable {
     # Note this is if initrd.network.enable, not if
     # initrd.systemd.network.enable. By setting the latter and not the
     # former, the user retains full control over the configuration.
     boot.initrd.systemd.network = mkMerge [(genericDhcpNetworks true) interfaceNetworks];
-  })];
+  })
 
-  config = mkIf cfg.useNetworkd {
+  (mkIf cfg.useNetworkd {
 
     assertions = [ {
       assertion = cfg.defaultGatewayWindowSize == null;
@@ -450,6 +452,7 @@ in
               bindsTo = [ "systemd-networkd.service" ];
           };
       };
-  };
+  })
 
+  ];
 }
```

Staying at the same indentation level as the `config` name is a little awkward but it seems the best option at the moment

lilyinstarlight replied:
That works for me too! And we could let the reindententation be picked up after we have an RFC101 blessed format/formatter for nixpkgs

lilyinstarlight commented on nixos/tests/predictable-interface-names.nix:
```diff
@@ -8,25 +8,48 @@ let
-        ip link
-        ip link show eth0 ${if predictable then "&&" else "||"} exit 1
-      '';
+      boot.initrd.postDeviceCommands = script;
+
+      boot.initrd.systemd = {
+        enable = systemdStage1;
+        initrdBin = [ pkgs.iproute2 ];
+        services.systemd-udev-settle.wantedBy = ["initrd.target"];
+        services.check-interfaces = {
+          requiredBy = ["initrd.target"];
+          after = ["systemd-udev-settle.service"];
+          serviceConfig.Type = "oneshot";
+          path = [ pkgs.iproute2 ];
+          inherit script;
+        };
+      };
```
Should setting these be conditional on `systemdStage1` instead of unconditionally set? (especially if we add asserts/warnings later on for the `boot.initrd.{pre,post}*Commands` options not supported with systemd stage-1...)

ElvishJerricco replied:
That's a fair point. It doesn't have any material effect, but might as well be written that way.

lilyinstarlight reviewed (commented):
This looks pretty fabulous; thank you for this work!

Most of the technical questions I would have had seem to have already been asked or addressed, but I did leave a few review comments

mlyxshi commented on nixos/modules/config/users-groups.nix:
```diff
@@ -639,10 +689,52 @@ in {
+    # systemd initrd
+    boot.initrd.systemd = lib.mkIf config.boot.initrd.systemd.enable {
+      contents = {
+        "/etc/passwd".text = ''
```
when initrd ssh on, the output is
```
-sh-5.2# cat /etc/passwd
messagebus:x:4:4::/var/empty:
nobody:x:65534:65534::/var/empty:
root:x:0:0::/var/empty:
sshd:x:1:1::/var/empty:
systemd-network:x:152:152::/var/empty:
```
I think root home directory should be `/root`, and the root shell should be `${boot.initrd.network.ssh.shell}`

ElvishJerricco replied:
@mlyxshi That would require having a means of configuring a user's shell and home directory in `boot.initrd.systemd.users`. Given that's a whole new feature, let's save that for a followup Pr.

lilyinstarlight commented on nixos/modules/services/hardware/udev.nix:
```diff
@@ -16,16 +16,6 @@ let
   '';
 
 
-  # networkd link files are used early by udev to set up interfaces early.
-  # This must be done in stage 1 to avoid race conditions between udev and
-  # network daemons.
-  # TODO move this into the initrd-network module when it exists
-  initrdLinkUnits = pkgs.runCommand "initrd-link-units" {} ''
-    mkdir -p $out
-    ln -s ${udev}/lib/systemd/network/*.link $out/
-    ${lib.concatMapStringsSep "\n" (file: "ln -s ${file} $out/") (lib.mapAttrsToList (n: v: "${v.unit}/${n}") (lib.filterAttrs (n: _: hasSuffix ".link" n) config.systemd.network.units))}
-  '';
-
```
Do these `config.systemd.network.units` link units still get copied to systemd-stage-1 anywhere?

I do not see some similar snippet added anywhere and generating an initrd with `boot.initrd.systemd.enable = true` and `systemd.network.links` (with `boot.initrd.network.enable` both on and off) does not seem to end up with any link units in `/etc/systemd/network` in initrd

ElvishJerricco replied:
@lilyinstarlight Good catch. I thought the `systemd.contents = mkUnitFiles "/etc/" cfg;` line was all we needed but that's clearly not quite the same thing since it isn't including all stage 2 link units.

ElvishJerricco replied:
@lilyinstarlight I *think* it's ok to just add this to `boot.initrd.systemd.network.units`, [like this](https://github.com/NixOS/nixpkgs/compare/0ff15452acae8164697dcb5bd7247d878ee59dad..3cb9534df64937f3425dbc0f63749cc83276791a).

flokli replied:
Can we extend / add one test to make sure the logic with an interface rename via .link is exercised?

Maybe also a .netdev file could be used to create a "dummy" interface, and ensure it's kept (probably easier to test than a vlan)

flokli replied:
I think it's a conceptual question on how much of the network config in initrd should match the one on the real filesystem.

Personally, I'd be in favor of copying all of it in - it's small text files, most of the time the config won't differ much from the real OS, and the user can always put a lexicographically smaller unit file before to override it specifically for in-initrd.

That way, enabling networkd in initrd will give you the same config, which is the one of the least surprises.

At least for the .link files we /need/ to do it though, as doing it after udev started and the interfaces are already brought up will be too late.

ElvishJerricco replied:
> Personally, I'd be in favor of copying all of it in

We definitely can't do that, particularly because of things like wireguard that need secret files and whatnot. We could copy in everything that doesn't need support files from the host FS, but I don't know if we have a way to categorically know which those are. Plus there are kernel modules to consider; it's not always clear which kernel modules would need to be added to initrd to support the entirety of someone's networkd config.

That said, you can (and I do) just do `boot.initrd.systemd.network = config.systemd.network` to make them the same.

Anyway, as for the link units, the PR that added this comment also [linked to an issue](https://github.com/systemd/systemd/issues/7293), but that issue was closed by [this](https://github.com/systemd/systemd/pull/11881) way *before* that PR was even made. I'm wondering why that systemd change isn't enough to resolve the problem. *Surely* you're not expected to do your interface renaming in stage 1. In fact, the comments in the issue suggest that the udev rules to rename the interface in stage 2 should happen before `networkd` is even notified the interface is ready.

lilyinstarlight replied:
> @lilyinstarlight I _think_ it's ok to just add this to `boot.initrd.systemd.network.units`, [like this](https://github.com/NixOS/nixpkgs/compare/0ff15452acae8164697dcb5bd7247d878ee59dad..3cb9534df64937f3425dbc0f63749cc83276791a). Does that seem good to you?

Yep! That diff looks good to me

> But again, that's probably a concern for a future PR.

I agree there, since it would otherwise be a behavior change and I'd love to get this in within the next day or two as-is

colemickens commented:
:confetti_ball: :rocket: I've been excited for this one for a while, thank you to everyone!
